patcher.sh

```diff
@@ -0,0 +1,302 @@
+		RESOURCES="${OUTPUT}/assets"
+	else
+		OUTPUT="${PREFIX}/am2r_${VERSION}"
+		RESOURCES="${OUTPUT}/assets"
+	fi
+	# Cleanup in case the dirs/files exists
+	cleanupDirs=("$OUTPUT" "$SCRIPT_DIR/AM2R.AppDir" "$SCRIPT_DIR/AM2R-x86_64.AppImage" "$SCRIPT_DIR/AM2RWrapper/" "$SCRIPT_DIR/utilities/android/assets/" "$SCRIPT_DIR/AM2RWrapper.apk")
+	rm -rf "${cleanupDirs[@]}"
+
+	mkdir -p "$OUTPUT"
+	mkdir -p "$RESOURCES"
+
+	# Check for AM2R_11
+	if [[ -f "$AM2RZIP" ]]; then
+		echo "AM2R_11.zip found!"
+		unzip -q "$AM2RZIP" -d "$OUTPUT"
+	elif [[ -d "$AM2RZIP" ]]; then
+		echo "AM2R_11 folder found!"
+		cp -R "$AM2RZIP" "$OUTPUT"
+	else
+		echo "AM2R_11 not found! Please place AM2R_11.zip (case sensitive) into \"$SCRIPT_DIR\" or provide it via command line arguments and try again."
+		return 1
+	fi
+
+	# Check for which OS we patch
+	if [ "$OSCHOICE" = "linux" ]; then
+
+		echo "Patching for Linux..."
+		echo "Applying AM2R xdelta patch..."
+		xdelta3 -dfs "$OUTPUT/AM2R.exe" "$SCRIPT_DIR/data/AM2R.xdelta" "$OUTPUT/runner"
+
+		echo "Applying data xdelta patch..."
+		xdelta3 -dfs "$OUTPUT/data.win" "$SCRIPT_DIR/data/game.xdelta" "$RESOURCES/game.unx"
+
+		echo "Cleaning up residual AM2R 1.1 files..."
+		rm "$OUTPUT/AM2R.exe" "$OUTPUT/data.win" "$OUTPUT/D3DX9_43.dll"
+
+		echo "Formatting game directory..."
+		# This won't move the runner, because doesn't have a dot
+		mv $OUTPUT/*.* "$RESOURCES/"
+
+		echo "Installing new datafiles..."
+		cp -R "$SCRIPT_DIR"/data/files_to_copy/* "$RESOURCES/"
+
+		if [ "$HQMUSIC" = true ]; then
+			echo "Copying HQ music..."
+			cp "$SCRIPT_DIR"/data/HDR_HQ_in-game_music/*.ogg "$RESOURCES/"
+		fi
+
+		# Format music for Unix, aka lowercase it
+		# Finds every file, and if its not already lowercase, lowercases it.
+		find "$RESOURCES" -type f -exec \
+			bash -c 'target="{}";
+				cd "$(dirname "${target}")";
+				target="$(basename "${target}")";
+				! [[ "${target}" = "${target,,}" ]] && mv "${target}" "${target,,}"' \;
+
+		# Remove old lang folder
+		rm -R "$OUTPUT"/lang
+
+		chmod +x "$OUTPUT/runner"
+
+		if [ "$SYSTEMWIDE" = true ]; then
+			mkdir "$PREFIX/bin"
+			ln -s "$OUTPUT/runner" "$PREFIX/bin/am2r"
+		fi
+
+		# Create .desktop file
+		echo "Creating desktop file..."
+		cp "$SCRIPT_DIR/data/files_to_copy/icon.png" "$OUTPUT/icon.png"
+		desktopPath="$OUTPUT/AM2R.desktop"
+		if [ "$SYSTEMWIDE" = true ]; then
+			mkdir -p "$PREFIX/share/applications"
+			desktopPath="$PREFIX/share/applications/AM2R.desktop"
+		fi
+		cp "$SCRIPT_DIR/DesktopTemplate" "$desktopPath"
+
+		# replace [REPLACE] with OUTPUT.Replace("/", "\/") in desktop file
+		sed -i "s/\[REPLACE\]/${OUTPUT//\//\\\/}/" "$desktopPath"
+
+		# If we don't want to create an AppImage, we're done.
+		if [ "$APPIMAGE" = false ]; then
+			# For non-appimage, the desktop file should point to runner
+			sed -i "s/AM2R.AppImage/runner/" "$desktopPath"
+			echo ""
+			echo "The operation was completed successfully. See you next mission!"
```

Comment:
Can be used later in the script for newlines as well.
```suggestion
			echo -n "\nThe operation was completed successfully. See you next mission!"
```

Reply:
Hm, prefer how this looks cleaner / easier to follow visible for most operations.
In here for example:
(everything nice indented)
```bash
	echo "Select your patch type:"
	echo ""
	echo "1 - Linux"
	echo "2 - Android"
	echo ""
	echo "Awaiting input:"
```
vs
(initial line is indented correct, following lines break indent)
```bash
echo -n "Select your patch type:

1 - Linux
2 - Android

Awaiting input:"
```
or even worse:
(single line)
```bash
echo -n "Select your patch type:\n\n1 - Linux\n2 - Android\n\nAwaiting input:"
```

Reply:
In my opinion the single line option does look the cleanest, but that's pretty subjective.

Reply:
Will leave this as-is then. IMO at least it helps with readability seeing the output exactly as it'll be shown when run, and I don't think any system this is targeted for would have problems due to multiple echo's.